python/tvm/driver/tvmc/runner.py

```diff
@@ -0,0 +1,450 @@
+
+    # First check all the keys in inputs exist in the graph
+    for input_name in inputs:
+        if input_name not in shape_dict.keys():
+            raise TVMCException("the input tensor '{}' is not in the graph".format(input_name))
+
+    # Now construct the input dict, generating tensors where no
+    # data already exists in 'inputs'
+    inputs_dict = {}
+    for input_name in shape_dict:
+        if input_name in inputs.keys():
+            logger.debug("setting input '%s' with user input data", input_name)
+            inputs_dict[input_name] = inputs[input_name]
+        else:
+            shape = shape_dict[input_name]
+            dtype = dtype_dict[input_name]
+
+            logger.debug(
+                "generating data for input '%s' (shape: %s, dtype: %s), using fill-mode '%s'",
+                input_name,
+                shape,
+                dtype,
+                fill_mode,
+            )
+            data = generate_tensor_data(shape, dtype, fill_mode)
+            inputs_dict[input_name] = data
+
+    return inputs_dict
+
+
+def run_module(
+    module_file,
+    hostname,
+    port=9090,
+    rpc_key=None,
+    device=None,
+    inputs=None,
+    fill_mode="zeros",
+    repeat=1,
+    profile=False,
+):
+    """Run a compiled graph runtime module locally or remotely with
+    optional input values.
+
+    If input tensors are not specified explicitly, they can be filled
+    with zeroes, ones or random data.
+
+    Parameters
+    ----------
+    module_file : str
+        The path to the module file (a .tar file).
+    hostname : str
+        The hostname of the target device on which to run.
+    port : int, optional
+        The port of the target device on which to run.
+    rpc_key : str, optional
+        The tracker key of the target device. If this is set, it
+        will be assumed that remote points to a tracker.
+    device: str, optional
+        the device (e.g. "cpu" or "gpu") to be targeted by the RPC
+        session, local or remote).
+    inputs : dict, optional
+        A dictionary of {input_name: np.array} storing the input
+        tensors to the network. If this is not specified, data will
+        be automatically generated for the input tensors.
+    fill_mode : str, optional
+        The fill-mode to use when generating data for input tensors.
+        Valid options are "zeros", "ones" and "random".
+        Defaults to "zeros".
+    repeat : int, optional
+        How many times to repeat the run.
+    profile : bool
+        Whether to profile the run with the debug runtime.
+
+    Returns
+    -------
+    outputs : dict
+        a dictionary with output tensors, generated by the module
+    times : list of str
+        execution times generated by the time evaluator
+    """
+    if not inputs:
+        inputs = {}
+
+    with tempfile.TemporaryDirectory() as tmp_dir:
+        logger.debug("extracting module file %s", module_file)
+        t = tarfile.open(module_file)
```

Comment:
let us support both mode (tar and .so). Because not all remote devices has compiler.

Reply:
When you say "both .tar and .so", is it clear that inside the .tar there is a `.so` (together with the `graph` and also the `params`? Can you clarify what is the difference?

Reply:
My meaning is when we compile, we could let users `lib.export_library`, which supports two formats: .tar and .so. If it is .tar, which has .o inside and will require remote devices extract tar files and call compiler to compile it into .so. If it is .so, remote devices only needs runtime and could load it directly, no need to extract tar files and has compiler.

Reply:
Oh, actually then you run `tvmc compile` we create that tar manually, with the `.so` inside. Then I think we are covered for that specific `.so` case.

https://github.com/apache/incubator-tvm/blob/master/python/tvm/driver/tvmc/compiler.py#L218

Let me know if there is something else I'm missing.

Reply:
Thanks for clarifing. The tar file is one collection, not the file of export_library